### PR TITLE
fix(tracing): trace waterfall + summary prompt + chat tab all lose data after each agent turn

### LINKED ISSUE #896: tracing: trace waterfall + summary prompt + chat tab all lose data after each agent turn
The trace UI (`/traces` in the TUI) drops most of the conversation data after every agent turn. Three observable symptoms in a normal multi-turn session on `altimate-default`:

1. **Waterfall view collapses to the system-prompt span after each agent turn.** Real session has many tool calls and generations; trace file on disk has 1 root span and `metadata: {}` once the turn completes.
2. **Summary tab shows "What was asked / No prompt recorded"** even when a real user prompt was typed.
3. **Chat tab shows only the last user message** in multi-turn sessions. Earlier user turns render as their assistant responses but the user's "▶ You" bubbles are missing.

## Repro

Open a TUI session, send a couple of messages with tool-using replies (e.g. "list directories", then "okay, how about patches"), open `/traces`, watch the file:

```bash
ls -la ~/.local/share/altimate-code/traces/ses_<id>.json
jq '{spans: (.spans | length), metadata, summary}' ~/.local/share/altimate-code/traces/ses_<id>.json
```

Before the fix: `spans: 1`, `metadata: {}`, summary counters at 0/null, regardless of how much activity the DB shows for the session. The viewer's chat tab shows only one `▶ You` bubble for the latest prompt.

## Scope

Closed by the fix in this PR. Filing for traceability so the merge cleanly closes a tracked symptom report.

### PR DESCRIPTION
**Pull request by sahrizvi**
Closes #896.

Follow-up to #867. Fixes the trace lifecycle and rendering bugs that survived #867's concurrency hardening because they live at a different layer of the system.

## What this fixes

| # | Bug | Fix |
|---|---|---|
| 1 | `session.status === "idle"` handler in `worker.ts` called `endTrace + sessionTraces.delete` after **every turn**. `idle` is a per-turn busy→idle transition, not session-end. Next event hit a cache miss, `Trace.create()` + `startTrace` snapshotted an empty 1-span file, clobbering the rich on-disk trace. | Handler removed. Trace lives as long as the worker has the session in `sessionTraces`. Finalization on shutdown + MAX_TRACES eviction only. |
| 2 | `getOrCreateTrace` cache-miss path never rehydrated from disk — worker restart / eviction always overwrote the existing trace. | New `Trace.rehydrateFromFile(sessionId)` restores spans / metadata / counters / clears root `endTime`. `getOrCreateTrace` prefers it before falling back to `startTrace`. |
| 3 | `worker.setWorkspace` was not idempotent → spurious calls cleared `sessionTraces`. | Track `currentWorkspaceID` at module scope; early-return on unchanged value. |
| 4 | `routes/session/index.tsx` workspaceID effect re-ran on every `session()` signal change. | `createEffect(on(() => session()?.workspaceID, ...))` so SolidJS dirty-tracks only `workspaceID`. |
| 5 | Prompt never captured. Three event handlers in `worker.ts` could feed `metadata.prompt`: (a) the user-role `message.updated` handler only fired if `info.summary.title \|\| info.summary.body` was populated, which doesn't happen on the first turn; (b) the text-part branch of `message.part.updated` gated on `part.time?.end`, which user-input parts never have set; (c) the `session.updated` handler called `setTitle(info.title)` with one argument only, never touching prompt. Codex review caught a second risk: dropping the `time.end` gate while keeping `setTitle(text, text)` would regress the auto-generated session title to raw user input, because `setTitle(t, p)` mutates both fields. | New `Trace.setPrompt(prompt)` that only mutates `metadata.prompt`. The user-text branch of `message.part.updated` drops the `time.end` precondition and calls `setPrompt(text)` instead of `setTitle(text, text)`. |
| 6 | Chat tab could only render `metadata.prompt` (singular) at the top + assistant generations. Every user message after the first was dropped from the viewer. | New `Trace.logUserMessage(text)` pushes a `kind: "user-message"` span. The user-text branch of `message.part.updated` also calls it for each user-identified text. Viewer chat-view rebuilt to interleave user-message and generation spans in `startTime` order. Backward compatible with traces that lack user-message spans. |

## Why #867 didn't catch these

PR #867 closed M2 (snapshot-debounce dropping events) and M3 (FileExporter racing flushSync). Both are **intra-Trace-instance concurrency** — within one Trace object's state machine. The reproducers in `tracing-rename-race.test.ts` and `tracing-display-crash.test.ts` all hold one instance for the test's lifetime and exercise crash/burst conditions against that instance.

The bugs in this PR are at three other layers:

- **Cross-instance lifecycle** (rows 1–4): when one Trace gets destroyed and a new one is constructed for the same `sessionID`. The on-disk file clobber happens via the *new* instance's first snapshot, not from any race within an old one.
- **Worker event-handler layer / data feed** (row 5): what gets fed into the Trace from the bus event stream. Orthogonal to persistence correctness.
- **Viewer renderer + data model** (row 6): how the viewer's chat tab iterates spans. A correct persistence layer can still ship a 1-span trace if upstream feeds 1 span.

Beyond the layer mismatch, the per-turn idle handler (row 1) was **intentional code** — the original comment read "Finalize trace when session reaches idle (completed)". It was a semantic misunderstanding of what `idle` means at the event level (per-turn vs session-end), not a concurrency bug, so #867's reproducer matrix wouldn't have surfaced it. And no existing test exercises the full event-stream-to-trace-file path across multiple `busy → idle` cycles, so the cross-turn nature was invisible to CI.

## Independent of #867's scope

Every bug fixed here predates #867 (the prior trace-corruption PR). Listing
this explicitly so reviewers can see this PR is orthogonal work — not a
follow-up cleaning up regressions #867 introduced.

| Bug fixed in this PR | Origin | Evidence |
|---|---|---|
| 1. Per-turn `session.status === "idle"` destroying the Trace (`worker.ts`) | Predates #867 | `git log -S 'status === "idle"' -- worker.ts` → introduced by #183 ("feat: replace Langfuse with local-first tracing system"), the original local-tracing PR. Predates #867 by months. |
| 2. `getOrCreateTrace` no disk rehydration (`worker.ts`) | Predates #867 | Same #183 PR introduced `getOrCreateTrace` without rehydration. #867 never touched `worker.ts`. |
| 3. `setWorkspace` not idempotent (`worker.ts`) | Predates #867 | Same as above — #867 never touched `worker.ts`. |
| 4. SolidJS effect re-firing on every signal change (`routes/session/index.tsx`) | Predates #867 | #867 never touched the TSX route at all. |
| 5. Prompt-capture broken (`worker.ts`) | Predates #867 | All three relevant event handlers (`message.updated` user-role branch, `message.part.updated` text branch, `session.updated`) live in `worker.ts`, never touched by #867. |
| 6. User messages not first-class in trace data model (`tracing.ts` kind union, `viewer.ts` chat-view) | Predates #867 | #867 modified `tracing.ts` but only added M2/M3 concurrency fixes — did NOT change the `kind` union, did NOT touch `viewer.ts` (zero changes to the viewer file in the #867 commits). |

## Tests

- `test/cli/tui/worker-trace-clearing.test.ts` — source-grep contract tests: no-idle-clobber, rehydrate-before-startTrace, `setPrompt`-not-`setTitle`, `Trace.setPrompt` purity.
- `test/altimate/tracing-rehydrate.test.ts` — behavioral tests for `rehydrateFromFile` (preserve spans/metadata/counters, reject missing/mismatched/malformed, clear endTime on rehydrate) and `logUserMessage` (chronological ordering).

47 affected tests pass. Typecheck clean. 35 existing tracing tests (`tracing-display-crash`, `tracing-rename-race`) pass unchanged.

## Verification on a live build

Built locally as `0.0.0-fix/trace-clearing-on-workspace-set-202606051512`. Confirmed against a real multi-turn TUI session that:
- The trace file now persists across `busy → idle` transitions (5+ spans, full metadata, real counter values where v0.8.3 produced a 1-span / `metadata: {}` / null-summary file).
- After a fresh rebuild with the rendering fix, multi-turn chat tab interleaves each user message with its assistant response in chronological order instead of showing only the last user prompt.

## Out of scope

- **Higher-fidelity integration test** for cross-turn trace correctness. The source-grep + behavioral tests here lock the contracts but don't drive synthetic `session.status` idle events through the real worker event loop. Tracked separately as a follow-up.
- The user-role `message.updated` handler in worker.ts still calls `setTitle(title, title)` with `summary.title || summary.body`. Same `setTitle(x, x)` shape Codex caught for the user-text branch in this PR; if the title-agent's summary update lands after `session.updated`, it can overwrite `metadata.prompt` with the title-agent's text instead of the user's actual input. Smaller surface than the bug fixed here; not addressed in this PR.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---
## Summary by cubic
Fixes trace lifecycle bugs that overwrote traces after each turn and preserves trace identity across restarts. Also makes the chat tab show multi-turn conversations, improves prompt capture (skipping synthetic parts), and prevents duplicate top-level prompts for long messages.

- **Bug Fixes**
  - Removed per-turn finalize on `session.status === "idle"`; traces persist across turns and only finalize on shutdown or eviction.
  - Added `Trace.rehydrateFromFile(sessionId)` and call it before `startTrace` on cache miss; preserves spans/metadata/counters/`traceId`, normalizes session ID, clears root `endTime`, and marks any open generation spans as “interrupted” to keep structure.
  - Made `setWorkspace` idempotent via `currentWorkspaceID`; UI effect now uses `on(() => session()?.workspaceID, ...)` so it fires only on real changes, preventing `sessionTraces` clearing.
  - Captured prompts via `Trace.setPrompt(prompt)` and dropped the `time.end` gate for user text; never call `setTitle(text, text)` and skip `synthetic`/`ignored` text parts so prompts stay accurate.
  - Logged user messages with `Trace.logUserMessage(text)` and updated the viewer to interleave `user-message` and `generation` spans by time; falls back to `metadata.prompt` when needed and dedupes using the shared `USER_MESSAGE_INPUT_MAX_CHARS` cap to avoid duplicate long prompts.

- **Refactors**
  - Made `Trace.rehydrateFromFile` async and awaited it in `getOrCreateTrace` and stream handlers to avoid blocking the event loop on cache misses; tests updated accordingly.
  - Migrated `tracing-rehydrate` tests to the `tmpdir()` fixture for consistency.
  - Tightened `worker-trace-clearing` test regex to be scope-bounded, ensuring the `synthetic/ignored` gate and `setPrompt`/`logUserMessage` checks apply to the correct branch.
  - Removed redundant `specs/trace-bugs-followup-867.md` to avoid duplicated documentation.

<sup>Written for commit 7f7115408871a33ecb786107b867f133d9080417. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/AltimateAI/altimate-code/pull/895?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

* **New Features**
  * Multi-turn chat with explicit per-turn user-message tracking and improved session recovery from disk

* **Bug Fixes**
  * Deduplicated prompt bubbles in the chat view and preserved in-progress traces across turns (prevents lost/overwritten history)
  * Reduced unnecessary workspace restarts to avoid clearing cached traces

* **Tests**
  * Added tests covering session rehydration, user-message recording, and trace-retention behaviors

### CODE REVIEW

**Comment by dev-punia-altimate** [comment]
## ✅ Tests — All Passed

### TypeScript — passed


cc @sahrizvi

File: packages/opencode/src/altimate/observability/tracing.ts [hunk 1]
```diff
@@ -490,6 +490,70 @@ export class Trace {
...
+    } catch {
+      return false
+    }
+    if (!trace || trace.sessionId !== sessionId || !Array.isArray(trace.spans) || trace.spans.length === 0) {
+      return false
+    }
+    const root = trace.spans.find((s) => s.parentSpanId === null && s.kind === "session")
+    if (!root) return false
+
+    this.sessionId = sessionId
```
[the remark this answers is not in the document]
Comment (sahrizvi) [reply packages/opencode/src/altimate/observability/tracing.ts:547]: Addressed in 38e0caed5. `rehydrateFromFile` now restores `this.traceId` from the parsed file (gated on a non-empty string so a malformed file can't blank it). Behavioral test in `tracing-rehydrate.test.ts` asserts the traceId on disk after a rehydrate+flush round-trip matches the original.

File: packages/opencode/src/altimate/observability/tracing.ts [hunk 2]
```diff
@@ -490,6 +490,70 @@ export class Trace {
...
+    } catch {
+      return false
+    }
+    let trace: TraceFile
+    try {
+      trace = JSON.parse(raw) as TraceFile
+    } catch {
+      return false
+    }
+    if (!trace || trace.sessionId !== sessionId || !Array.isArray(trace.spans) || trace.spans.length === 0) {
```
[the remark this answers is not in the document]
Comment (sahrizvi) [reply packages/opencode/src/altimate/observability/tracing.ts]: Addressed in 38e0caed5. The comparison now normalizes `sessionId` via the same `.replace(/[/\\\\.:]/g, "_")` that `buildTraceFile` applies before writing (the file we're reading was written with the sanitized form). Behavioral test in `tracing-rehydrate.test.ts` calls `rehydrateFromFile("ses_with/slash.and:colon")` against a file written with the sanitized session id — pre-fix this returned false.

File: packages/opencode/src/altimate/observability/viewer.ts [hunk 3]
```diff
@@ -1310,13 +1310,29 @@ function showDetail(span) {
...
   var html = '';
-  if (t.metadata.prompt) {
+  // Build a chronologically sorted list of conversation turns by interleaving
+  // user-message spans with generation spans. Older traces (no user-message
+  // spans) fall back to rendering metadata.prompt at the top as a single "You"
+  // bubble. Once user-message spans are present, we prefer those and skip
+  // metadata.prompt to avoid double-rendering the first turn.
+  var userMsgs = spans.filter(function(s){return s.kind==='user-message';});
+  var gens = spans.filter(function(s){return s.kind==='generation';});
+  if (userMsgs.length === 0 && t.metadata.prompt) {
```
[the remark this answers is not in the document]
Comment (sahrizvi) [reply packages/opencode/src/altimate/observability/viewer.ts]: Addressed in 38e0caed5. The fallback no longer hard-gates on `userMsgs.length === 0`. New logic: render `metadata.prompt` at the top unless some existing user-message span already carries the same text. Three cases now work — pre-fix traces (no user-message spans), mixed traces (rehydrated pre-fix data + new turns), and new traces (avoids duplicating the first turn that's already in a span).

File: packages/opencode/test/cli/tui/worker-trace-clearing.test.ts [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+    )
+
+    // The previous shape (`createEffect(() => { if (session()?.workspaceID) ... })`)
+    // re-runs on every session() signal change. The fixed shape uses `on()` with
+    // an explicit projector returning workspaceID, so SolidJS dirty-tracks only
+    // that value.
+    expect(routeSrc).toMatch(/createEffect\(\s*on\(\s*\(\)\s*=>\s*session\(\)\?\.workspaceID/)
+
+    // The unguarded inline form must not reappear — that would re-introduce the bug.
+    expect(routeSrc).not.toMatch(/createEffect\(\(\)\s*=>\s*\{\s*\n\s*if\s*\(session\(\)\?\.workspaceID\)/)
```
[the remark this answers is not in the document]
Comment (sahrizvi) [reply packages/opencode/test/cli/tui/worker-trace-clearing.test.ts]: Addressed in 38e0caed5. The two guards now cover the flagged shapes:
- workspaceID effect: rejects inline-AND (`createEffect(() => session()?.workspaceID && ...)`), inline-ternary (`createEffect(() => session()?.workspaceID ? ... : ...)`), and block-body with `if` — three separate negative regexes, each bounded with `[^{}]` for the block form so the pattern can't cross unrelated createEffects in the same file (the route file has several).
- user-text part: in addition to the existing strict `if (part.type === "text" && part.time?.end)` rejection, a broader guard now rejects any `part.time?.end` check appearing inside the user-text branch (gated on `sessionUserMsgIds.get(part.sessionID)?.has(part.messageID)`).

Caveat I bumped into and fixed: the broadened workspaceID regex first matched my own comment in `routes/session/index.tsx` that quoted the bug shape literally — paraphrased the comment to describe the behavior without reproducing the source line.

[the remark this answers is not in the document]
Comment (sahrizvi) [reply packages/opencode/test/cli/tui/worker-trace-clearing.test.ts]: Folded into the response on the parallel CodeRabbit thread — same broadening landed in 38e0caed5.

File: packages/opencode/test/altimate/tracing-rehydrate.test.ts [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+let tmpDir: string
+
+beforeEach(async () => {
+  tmpDir = path.join(os.tmpdir(), `tracing-rehydrate-${Date.now()}-${Math.random().toString(36).slice(2)}`)
+  await fs.mkdir(tmpDir, { recursive: true })
+})
+
+afterEach(async () => {
+  await fs.rm(tmpDir, { recursive: true, force: true }).catch(() => {})
+})
```
[the remark this answers is not in the document]
Comment (sahrizvi) [reply packages/opencode/test/altimate/tracing-rehydrate.test.ts]: Skipping. The `tmpdir()` fixture helper is a sensible convention, but the existing tracing tests in this directory (`tracing-display-crash`, `tracing-rename-race`, `tracing-adversarial-*`) all use the same manual `os.tmpdir() + beforeEach/afterEach` pattern this file uses. Migrating one file alone would be inconsistent; a sweep across the directory is out of scope for this PR. Worth a separate cleanup PR if the project wants to standardize.

[the remark this answers is not in the document]
Comment (sahrizvi) [reply packages/opencode/test/altimate/tracing-rehydrate.test.ts]: Reversed my earlier decision after seeing CodeRabbit's reply citing `packages/opencode/test/AGENTS.md` — the convention IS documented, the sibling tests just predate it. For code I'm adding fresh, following the standard is correct.

Done in 7bee91372: `tracing-rehydrate.test.ts` now uses `await using tmp = await tmpdir()` per test, with helpers `makeTrace(dir)` and `readTraceFile(dir, sessionId)` threading the tmp path through. Manual `beforeEach`/`afterEach` removed. 46 affected tests still pass.

The dir-wide sweep (`tracing-display-crash`, `tracing-rename-race`, `tracing-adversarial-*`) remains out of scope for this PR — happy for that to be a follow-up.

File: packages/opencode/src/altimate/observability/viewer.ts [hunk 6]
```diff
@@ -1310,13 +1310,39 @@ function showDetail(span) {
...
+  //      later turns. Without this, the legacy first turn would be dropped.
+  //   3. Brand-new traces where the first user-message span input equals
+  //      metadata.prompt — we skip the duplicate to avoid double-rendering.
+  var userMsgs = spans.filter(function(s){return s.kind==='user-message';});
+  var gens = spans.filter(function(s){return s.kind==='generation';});
   if (t.metadata.prompt) {
-    html += '<div class="chat-msg user"><div class="chat-role">\\u25B6 You</div>';
-    html += '<div class="chat-bubble">' + e(t.metadata.prompt) + '</div></div>';
+    var promptAlreadyInSpan = userMsgs.some(function(u){
+      return typeof u.input === 'string' && u.input === t.metadata.prompt;
```
[the remark this answers is not in the document]
Comment (sahrizvi) [reply packages/opencode/src/altimate/observability/viewer.ts]: Addressed in b110980f1. The dedupe now matches `metadata.prompt` against both the full string and its `.slice(0, 4000)` form — same fix shape you suggested. Prompts longer than the truncation boundary no longer render twice.

File: packages/opencode/test/altimate/tracing-rehydrate.test.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (sahrizvi) [reply packages/opencode/test/altimate/tracing-rehydrate.test.ts]: Updating the thread: I reversed the skip after re-reading your reply and migrated `tracing-rehydrate.test.ts` in 7bee91372 — that file now uses `await using tmp = await tmpdir()` per test. The cached learning from this thread is therefore slightly off for that one file (it generalizes from sibling tests that haven't been swept yet).

No need to file a follow-up issue from your side — the dir-wide sweep for the other tracing tests is the kind of cleanup the repo owners should scope themselves.

File: packages/opencode/src/altimate/observability/viewer.ts [hunk 7]
```diff
@@ -1310,13 +1310,45 @@ function showDetail(span) {
...
+  var gens = spans.filter(function(s){return s.kind==='generation';});
   if (t.metadata.prompt) {
-    html += '<div class="chat-msg user"><div class="chat-role">\\u25B6 You</div>';
-    html += '<div class="chat-bubble">' + e(t.metadata.prompt) + '</div></div>';
+    // \`logUserMessage\` truncates span input to 4000 chars (see tracing.ts);
+    // \`metadata.prompt\` stores the full string. For prompts longer than the
+    // truncation length, strict equality would miss the dedupe and the same
+    // text would render twice. Match against the truncated form as well.
+    var promptStr = String(t.metadata.prompt);
+    var promptTruncated = promptStr.slice(0, 4000);
```
[the remark this answers is not in the document]
Comment (sahrizvi) [reply packages/opencode/src/altimate/observability/viewer.ts]: Addressed in 39dd742e2. Extracted the 4000-char cap as `USER_MESSAGE_INPUT_MAX_CHARS` (exported from `tracing.ts`). `logUserMessage` reads it directly; the viewer's chat-tab dedupe interpolates it at render time so the JS and TS sides can't drift if the cap ever changes. Same commit also lands the codex-validated Major #1 (synthetic-part gate on the user-text branch) and Major #2 (close in-flight generation spans as interrupted on rehydrate) from the multi-LLM consensus review.

**Review by dev-punia-altimate — COMMENTED** [review]
### Multi-Persona Review — Verdict: **ship**

The PR implements a well-structured feature to add a Cluster column and enhanced filtering to Databricks Job Run History, following React best practices, API conventions, and defensive rendering patterns. All adversarial tests passed, no critical or high-severity findings were identified, and the codebase hygiene is clean.

**15/15 agents completed** · 249s · 5 findings (0 critical, 0 high, 0 medium)

#### Low
- **[web-researcher]** PR handles cluster_id as string to preserve leading zeros and avoid precision loss, aligning with frontend best practices for non-numeric identifiers. → `apps/web/src/helpers/apis.ts:5712`
  - 💡 Consider adding a type annotation for cluster_id as string in the API contract for clarity.
- **[web-researcher]** Defensive rendering of null/undefined cluster_name with fallbacks ('-' or cluster_id) and data-testid mitigates potential XSS risks from untrusted data. → `apps/web/src/pages/jobs/components/JobRunHistory.tsx:113`
  - 💡 Consider adding a unit test that explicitly verifies the fallback rendering of null/undefined cluster_name.
- **[web-researcher]** Use of @altimateai/lego v2.1.0 Stack component with row layout and justify='between' leverages recent accessibility improvements in the design system. → `apps/web/src/components/tables/DataTableHeaderPortals.tsx:25`
  - 💡 Verify that the Stack component's updated spacing behavior is visually consistent across all breakpoints.
- **[context-pack]** Extracted DataTableHeaderPortals component reduces duplication between JobsTable and JobRunHistory, improving maintainability and consistency. → `apps/web/src/components/tables/DataTableHeaderPortals.tsx:1`
  - 💡 Consider documenting this component in the design system wiki for future reuse.
- **[web-researcher]** Adding result_states and cluster_ids query parameters aligns with Databricks API conventions for efficient server-side filtering. → `apps/web/src/helpers/apis.ts:5714`
  - 💡 Ensure backend PR AltimateAI/altimate-backend#5455 is merged in sync to avoid API mismatch.

---
<sub>Multi-Persona Review · vllm:qwen3-next-80b (waves) + vllm-fallback (synth) · </sub>

File: packages/opencode/test/cli/tui/worker-trace-clearing.test.ts [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+  test("user-text branch skips synthetic/ignored parts before calling setPrompt+logUserMessage", async () => {
+    const workerSrc = await fs.readFile(path.join(ROOT, "src/cli/cmd/tui/worker.ts"), "utf-8")
+    // The synthetic+ignored gate must exist literally somewhere.
+    expect(workerSrc).toMatch(/!part\.synthetic\s*&&\s*!part\.ignored/)
+    // And the two write paths must be co-located inside a guard built from the
+    // same predicate — verified by checking that the `isAuthoredText &&` gate
+    // appears, followed (lazily) by both `trace.setPrompt` and `trace.logUserMessage`.
+    expect(workerSrc).toMatch(
+      /isAuthoredText\s*&&[\s\S]*?trace\.setPrompt[\s\S]*?trace\.logUserMessage/,
+    )
```
[the remark this answers is not in the document]
Comment (sahrizvi) [reply packages/opencode/test/cli/tui/worker-trace-clearing.test.ts]: Addressed in 6267ec56f. The two assertions are now tightened:

1. The `isAuthoredText` declaration check requires the const to be built from both flags: `const isAuthoredText = !part.synthetic && !part.ignored`. Stronger than searching for the literal anywhere in the file.
2. The two write-path checks anchor on the `sessionUserMsgIds.get(part.sessionID)?.has(part.messageID)` branch and require both `trace.setPrompt` and `trace.logUserMessage` to live inside the same `if (text) { ... }` body. The `[^{}]` bounds on the inner spans prevent the match from crossing the closing brace of that body — same defensive shape we used for the `createEffect` regex in this file.

Didn't use your exact diff because my actual code declares `isAuthoredText` separately rather than inlining `!part.synthetic && !part.ignored` in the if condition — your regex assumed the inline form. The two patterns I shipped enforce the same end-to-end constraint without forcing a particular condition shape.

**Comment by sahrizvi** [comment]
@anandgupta42 Ready to be merged.

File: packages/opencode/src/altimate/observability/tracing.ts [hunk 9]
```diff
@@ -490,6 +496,103 @@ export class Trace {
...
+  //
+  // Does NOT call snapshot — the file is already correct on disk; an
+  // unnecessary write here would compete with concurrent flushSync paths.
+  rehydrateFromFile(sessionId: string): boolean {
+    if (!this.snapshotDir) return false
+    const safeId = sessionId.replace(/[/\\.:]/g, "_") || "unknown"
+    const filePath = path.join(this.snapshotDir, `${safeId}.json`)
+    let raw: string
+    try {
+      raw = fsSync.readFileSync(filePath, "utf-8")
```
[the remark this answers is not in the document]
Comment (sahrizvi) [reply packages/opencode/src/altimate/observability/tracing.ts]: Addressed in 7f7115408. `Trace.rehydrateFromFile` is now async (uses `fs.readFile` instead of `fsSync.readFileSync`), `getOrCreateTrace` in worker.ts is async too, and the three call sites in the event-stream loop now `await` it. The loop body was already `for await`, so the conversion stays local. 15 affected behavioural + source-grep tests pass.

File: packages/opencode/src/cli/cmd/tui/worker.ts [hunk 10]
```diff
@@ -203,16 +211,50 @@ const startEventStream = (input: { directory: string; workspaceID?: string }) =>
...
+                  // and ignored flags exist precisely to mark non-authored content; this
+                  // is exactly the place to consult them. We skip silently rather than
+                  // `continue`-ing the event-loop iteration because the outer loop still
+                  // needs to forward the event downstream via `Rpc.emit`.
+                  const isAuthoredText = !part.synthetic && !part.ignored
+                  // altimate_change end
+                  if (
+                    isAuthoredText &&
+                    part.messageID &&
+                    sessionUserMsgIds.get(part.sessionID)?.has(part.messageID)
```
[the remark this answers is not in the document]
Comment (sahrizvi) [reply packages/opencode/src/cli/cmd/tui/worker.ts:242]: Real race in the abstract, but bounded in practice. Producer side: `Session.createUserMessage` in `session/prompt.ts` calls `Session.updateMessage` (which emits `message.updated`) BEFORE `Session.updatePart` (which emits `message.part.updated`) for each text part. Consumer side: the worker reads `for await (const event of events.stream)` — events are processed in receipt order from a single stream. So as long as the producer ordering holds, `sessionUserMsgIds` is populated before the part-handler fires.

Not closing the door entirely — if the producer ever reorders (parallel batched emit, retry-on-reconnect, server-side reordering), this would manifest as silently dropped user text. The defensive shape would be to buffer parts by messageID until we see a `message.updated` that classifies them, then flush. I'd rather land that as a focused follow-up with a real reproducer than speculate the implementation here — happy to open an issue if you'd like.
